code/modules/projectiles/guns/energy/hos.dm

```diff
@@ -0,0 +1,273 @@
+
+	for(var/index in 1 to stored_ammo.len)
+		var/true_index = ((our_slot + index - 1) % stored_ammo.len) + 1 // Stupid ONE BASED lists!
+		var/obj/item/ammo_casing/mws_batt/next_batt = stored_ammo[true_index]
+		if(chambered != next_batt && !istype(next_batt, chambered.type) && next_batt.cell.charge >= next_batt.e_cost)
+			switch_to(next_batt, user)
+			break
+
+/obj/item/gun/ballistic/revolver/mws/AltClick(mob/living/user)
+	.=..()
+	if(magazine)
+		user.put_in_hands(magazine)
+		magazine.update_overlays()
+		if(magazine.ammo_count())
+			playsound(src, 'sound/weapons/gun_magazine_remove_full.ogg', 70, 1)
+		else
+			playsound(src, "gun_remove_empty_magazine", 70, 1)
+		magazine = null
+		to_chat(user, "<span class='notice'>You pull the magazine out of [src].</span>")
+		if(chambered)
+			chambered = null
+		update_overlays()
+
+/obj/item/gun/ballistic/revolver/mws/update_overlays()
+	.=..()
+	cut_overlays()
+	if(!chambered)
+		return
+
+	var/obj/item/ammo_casing/mws_batt/batt = chambered
+	var/batt_color = batt.type_color //Used many times
+
+	//Mode bar
+	var/image/mode_bar = image(icon, icon_state = "[initial(icon_state)]_type")
+	mode_bar.color = batt_color
+	add_overlay(mode_bar)
+
+	//Barrel color
+	var/image/barrel_color = image(icon, icon_state = "[initial(icon_state)]_barrel")
+	barrel_color.alpha = 150
+	barrel_color.color = batt_color
+	add_overlay(barrel_color)
+
+	//Charge bar
+	var/ratio = can_shoot() ? CEILING(clamp(batt.cell.charge / batt.cell.maxcharge, 0, 1) * charge_sections, 1) : 0
+	for(var/i = 0, i < ratio, i++)
+		var/image/charge_bar = image(icon, icon_state = "[initial(icon_state)]_charge")
+		charge_bar.pixel_x = i
+		charge_bar.color = batt_color
+		add_overlay(charge_bar)
+
+// The Magazine //
+/obj/item/ammo_box/magazine/mws_mag
+	name = "microbattery magazine"
+	desc = "A microbattery holder for the 'Big Iron'"
+
+	icon = 'icons/obj/ammo.dmi'
+	icon_state = "mws_mag"
+	caliber = "mws"
+	ammo_type = /obj/item/ammo_casing/mws_batt
+	start_empty = TRUE
+	max_ammo = 3
+
+	var/list/modes = list()
+
+/obj/item/ammo_box/magazine/mws_mag/update_overlays()
+	.=..()
+	cut_overlays()
+	if(!stored_ammo.len)
+		return //Why bother
+
+	var/x_offset = 5
+	var/current = 0
+	for(var/B in stored_ammo)
+		var/obj/item/ammo_casing/mws_batt/batt = B
+		var/image/cap = image(icon, icon_state = "[initial(icon_state)]_cap")
+		cap.color = batt.type_color
+		cap.pixel_x = current * x_offset //Caps don't need a pixel_y offset
+		add_overlay(cap)
+		if(batt.cell.charge > 0)
+			var/ratio = CEILING(clamp(batt.cell.charge / batt.cell.maxcharge, 0, 1) * 4, 1) //4 is how many lights we have a sprite for
+			var/image/charge = image(icon, icon_state = "[initial(icon_state)]_charge-[ratio]")
+			charge.color = "#29EAF4" //Could use battery color but eh.
+			charge.pixel_x = current * x_offset
+			add_overlay(charge)
+
+		current++ //Increment for offsets
+
+// The Casing //
+/obj/item/ammo_casing/mws_batt
```

Comment:
Move this to `code/modules/projectiles/ammunition/ballistic/revolver.dm` or something.